File: crates/flighthq-camera/src/shadow_camera.rs
//! `setup_directional_shadow_camera` — configures a camera for directional shadow rendering.

use flighthq_types::{Aabb, Camera, Vector3};

use crate::camera::set_camera_view_matrix4_from_look_at;
use crate::projection::create_orthographic_projection;

/// Configures `camera` as the shadow camera for a directional light.
///
/// Looks along `light_direction` (the light's travel direction) at the centre of
/// `scene_bounds`, with an orthographic frustum sized to the scene's bounding
/// sphere so the whole scene fits the shadow map. A degenerate (single-point)
/// bounds falls back to unit radius so the frustum stays valid.
///
/// Backend-agnostic CPU math (no GPU). `light_direction` need not be normalized.
/// The render side then draws scene depth from this camera into the shadow map and
/// samples it during shading.
pub fn setup_directional_shadow_camera(
    camera: &mut Camera,
    light_direction: &Vector3,
    scene_bounds: &Aabb,
) {
    let min = &scene_bounds.min;
    let max = &scene_bounds.max;
    let cx = (min.x + max.x) * 0.5;
    let cy = (min.y + max.y) * 0.5;
    let cz = (min.z + max.z) * 0.5;

    let mut radius = f32::hypot(f32::hypot(max.x - cx, max.y - cy), max.z - cz);
    if radius == 0.0 {
        radius = 1.0;
    }

    let dl = f32::sqrt(
        light_direction.x * light_direction.x
            + light_direction.y * light_direction.y
            + light_direction.z * light_direction.z,
    );
    let dl = if dl == 0.0 { 1.0 } else { dl };
    let dx = light_direction.x / dl;
    let dy = light_direction.y / dl;
    let dz = light_direction.z / dl;

    let distance = radius * 2.0;
    let eye = Vector3 {
        x: cx - dx * distance,
        y: cy - dy * distance,
        z: cz - dz * distance,
    };
    let target = Vector3 {
        x: cx,
        y: cy,
        z: cz,
    };
    // A near-vertical light needs a non-parallel up vector.
    let up = if dy.abs() > 0.99 {
        Vector3 {
            x: 0.0,
            y: 0.0,
            z: 1.0,
        }
    } else {
        Vector3 {
            x: 0.0,
            y: 1.0,
            z: 0.0,
        }
    };

    set_camera_view_matrix4_from_look_at(camera, &eye, &target, &up);
    camera.near = radius;
    camera.far = radius * 3.0;
    camera.projection = create_orthographic_projection(radius, radius);
}

#[cfg(test)]
mod tests {
    use flighthq_types::{Aabb, Vector3};

    use super::*;
    use crate::camera::create_camera;
    use crate::projection::create_perspective_projection;

    // setup_directional_shadow_camera

    #[test]
    fn setup_directional_shadow_camera_sets_orthographic_projection() {
        let mut camera = create_camera(0.1, 100.0, create_perspective_projection(1.0, 1.0));
        let bounds = Aabb {
            min: Vector3 {
                x: -5.0,
                y: -5.0,
                z: -5.0,
            },
            max: Vector3 {
                x: 5.0,
                y: 5.0,
                z: 5.0,
            },
        };
        let light_dir = Vector3 {
            x: 0.0,
            y: -1.0,
            z: 0.0,
        };

        setup_directional_shadow_camera(&mut camera, &light_dir, &bounds);

        // Orthographic frustum should be set to the bounding sphere radius.
        match &camera.projection {
            flighthq_types::Projection::Orthographic(_) => {}
            _ => panic!("expected orthographic projection"),
        }
    }

    #[test]
    fn setup_directional_shadow_camera_degenerate_bounds_uses_unit_radius() {
        let mut camera = create_camera(0.1, 100.0, create_perspective_projection(1.0, 1.0));
        let point = Aabb {
            min: Vector3 {
                x: 0.0,
                y: 0.0,
                z: 0.0,
            },
            max: Vector3 {
                x: 0.0,
                y: 0.0,
                z: 0.0,
            },
        };
        let light_dir = Vector3 {
            x: 0.0,
            y: -1.0,
            z: 0.0,
        };

        setup_directional_shadow_camera(&mut camera, &light_dir, &point);

        // Degenerate: radius falls back to 1. near = 1, far = 3.
        assert!((camera.near - 1.0).abs() < 1e-5);
        assert!((camera.far - 3.0).abs() < 1e-5);
    }

    #[test]
    fn setup_directional_shadow_camera_near_vertical_light_uses_z_up() {
        let mut camera = create_camera(0.1, 100.0, create_perspective_projection(1.0, 1.0));
        let bounds = Aabb {
            min: Vector3 {
                x: -1.0,
                y: -1.0,
                z: -1.0,
            },
            max: Vector3 {
                x: 1.0,
                y: 1.0,
                z: 1.0,
            },
        };
        // Near-vertical light (dy > 0.99 threshold)
        let light_dir = Vector3 {
            x: 0.0,
            y: -1.0,
            z: 0.0,
        };
        // Should not panic even with nearly-parallel up vector.
        setup_directional_shadow_camera(&mut camera, &light_dir, &bounds);
        assert!(camera.far > camera.near);
    }

    #[test]
    fn setup_directional_shadow_camera_unnormalized_direction_is_accepted() {
        let mut camera = create_camera(0.1, 100.0, create_perspective_projection(1.0, 1.0));
        let bounds = Aabb {
            min: Vector3 {
                x: -2.0,
                y: -2.0,
                z: -2.0,
            },
            max: Vector3 {
                x: 2.0,
                y: 2.0,
                z: 2.0,
            },
        };
        // Un-normalized (length 5)
        let light_dir = Vector3 {
            x: 0.0,
            y: 0.0,
            z: 5.0,
        };
        setup_directional_shadow_camera(&mut camera, &light_dir, &bounds);
        // far should be about 3× the bounding sphere radius of this 4x4x4 cube
        assert!(camera.far > 0.0);
    }
}

File: packages/camera/src/shadowCamera.ts
import type { AabbLike, Camera3D, Vector3Like } from '@flighthq/types';

import { setCamera3DViewMatrix4FromLookAt } from './camera';
import { createOrthographicProjection } from './projection';

// Configures `camera` as the shadow camera for a directional light — the spec's "a shadow camera is
// just a Camera3D placed at the light." Looks along `lightDirection` (the light's travel direction) at
// the centre of `sceneBounds`, with an orthographic frustum sized to the scene's bounding sphere so the
// whole scene fits the shadow map. The render side then draws scene depth from this camera into the
// shadow map and samples it during shading.
//
// Backend-agnostic CPU math (no GL). `lightDirection` need not be normalized. A degenerate (single
// point) bounds falls back to unit radius so the frustum stays valid.
export function configureDirectionalShadowCamera3D(
  camera: Camera3D,
  lightDirection: Readonly<Vector3Like>,
  sceneBounds: Readonly<AabbLike>,
): void {
  const min = sceneBounds.min;
  const max = sceneBounds.max;
  const cx = (min.x + max.x) * 0.5;
  const cy = (min.y + max.y) * 0.5;
  const cz = (min.z + max.z) * 0.5;

  let radius = Math.hypot(max.x - cx, max.y - cy, max.z - cz);
  if (radius === 0) radius = 1;

  const dl = Math.hypot(lightDirection.x, lightDirection.y, lightDirection.z) || 1;
  const dx = lightDirection.x / dl;
  const dy = lightDirection.y / dl;
  const dz = lightDirection.z / dl;

  // Pull the eye back two radii along -lightDirection so the whole sphere sits in front of the camera.
  const distance = radius * 2;
  _eye.x = cx - dx * distance;
  _eye.y = cy - dy * distance;
  _eye.z = cz - dz * distance;
  _target.x = cx;
  _target.y = cy;
  _target.z = cz;
  // A near-vertical light needs a non-parallel up vector.
  const up = Math.abs(dy) > 0.99 ? _upZ : _upY;

  setCamera3DViewMatrix4FromLookAt(camera, _eye, _target, up);
  camera.near = radius; // distance - radius
  camera.far = radius * 3; // distance + radius
  camera.projection = createOrthographicProjection({ halfHeight: radius, halfWidth: radius });
}

// Configures a directional-light shadow camera with a tight light-space fit of `worldBounds`.
// Unlike configureDirectionalShadowCamera3D's rotation-stable bounding sphere, this transforms all
// eight world-AABB corners into the light view and independently fits the orthographic X/Y extents,
// preserving substantially more texel density for a large static architectural scene. `padding`
// multiplies every fitted half-extent (default 1); callers can raise it to absorb animated pose
// excursions outside bind-pose aggregate bounds.
//
// A tight fit of static whole-scene world bounds is stable. Re-fitting per-frame visible bounds
// requires texel snapping to avoid shimmering and is intentionally a separate, caller-composed pass.
export function configureDirectionalShadowCamera3DTightFit(
  camera: Camera3D,
  lightDirection: Readonly<Vector3Like>,
  worldBounds: Readonly<AabbLike>,
  padding = 1,
): void {
  const min = worldBounds.min;
  const max = worldBounds.max;
  const hasBounds = min.x <= max.x && min.y <= max.y && min.z <= max.z;
  const cx = hasBounds ? (min.x + max.x) * 0.5 : 0;
  const cy = hasBounds ? (min.y + max.y) * 0.5 : 0;
  const cz = hasBounds ? (min.z + max.z) * 0.5 : 0;
  const ex = hasBounds ? (max.x - min.x) * 0.5 : 1;
  const ey = hasBounds ? (max.y - min.y) * 0.5 : 1;
  const ez = hasBounds ? (max.z - min.z) * 0.5 : 1;
  const radius = Math.max(Math.hypot(ex, ey, ez), 1e-4);
  const extentScale = padding > 0 ? padding : 1;

  let dx = lightDirection.x;
  let dy = lightDirection.y;
  let dz = lightDirection.z;
  const directionLength = Math.hypot(dx, dy, dz);
  if (directionLength > 0) {
    dx /= directionLength;
    dy /= directionLength;
    dz /= directionLength;
  } else {
    dx = 0;
    dy = -1;
    dz = 0;
  }

  // Pull back far enough that even padded depth extents remain in front of the camera. Translation
  // along the view direction does not change the tight X/Y fit.
  const distance = Math.max(radius * 2 * extentScale, 1);
  _eye.x = cx - dx * distance;
  _eye.y = cy - dy * distance;
  _eye.z = cz - dz * distance;
  _target.x = cx;
  _target.y = cy;
  _target.z = cz;
  setCamera3DViewMatrix4FromLookAt(camera, _eye, _target, Math.abs(dy) > 0.99 ? _upZ : _upY);

  const view = camera.view.m;
  let halfWidth = 0;
  let halfHeight = 0;
  let minViewZ = Number.POSITIVE_INFINITY;
  let maxViewZ = Number.NEGATIVE_INFINITY;
  for (let corner = 0; corner < 8; corner++) {
    const x = (corner & 1) === 0 ? cx - ex : cx + ex;
    const y = (corner & 2) === 0 ? cy - ey : cy + ey;
    const z = (corner & 4) === 0 ? cz - ez : cz + ez;
    const viewX = view[0] * x + view[4] * y + view[8] * z + view[12];
    const viewY = view[1] * x + view[5] * y + view[9] * z + view[13];
    const viewZ = view[2] * x + view[6] * y + view[10] * z + view[14];
    halfWidth = Math.max(halfWidth, Math.abs(viewX));
    halfHeight = Math.max(halfHeight, Math.abs(viewY));
    minViewZ = Math.min(minViewZ, viewZ);
    maxViewZ = Math.max(maxViewZ, viewZ);
  }

  const depthCenter = (minViewZ + maxViewZ) * 0.5;
  const halfDepth = Math.max((maxViewZ - minViewZ) * 0.5 * extentScale, 1e-4);
  camera.near = Math.max(-depthCenter - halfDepth, 1e-4);
  camera.far = Math.max(-depthCenter + halfDepth, camera.near + 1e-4);
  camera.projection = createOrthographicProjection({
    halfHeight: Math.max(halfHeight * extentScale, 1e-4),
    halfWidth: Math.max(halfWidth * extentScale, 1e-4),
  });
}

const _eye: Vector3Like = { x: 0, y: 0, z: 0 };
const _target: Vector3Like = { x: 0, y: 0, z: 0 };
const _upY: Vector3Like = { x: 0, y: 1, z: 0 };
const _upZ: Vector3Like = { x: 0, y: 0, z: 1 };
